File: Data Structures & Algorithms/longest-increasing-subsequence/submission-3.cpp
class Solution {
public:
int solve(int index,int prevIndex,vector<int>& nums,vector<vector<int>>& dp){
    if(index==nums.size()) return 0;
    if(dp[index][prevIndex+1]!=-1) return dp[index][prevIndex+1];
    int skip=solve(index+1,prevIndex,nums,dp);
    int take=0;
    if(prevIndex==-1 || nums[index]>nums[prevIndex])
    take=1+solve(index+1,index,nums,dp);
    return dp[index][prevIndex+1]=max(take,skip);
}
    int lengthOfLIS(vector<int>& nums) {
        int n=nums.size();
        vector<vector<int>> dp(n,vector<int>(n+1,-1));
        return solve(0,-1,nums,dp);
    }
};

File: Data Structures & Algorithms/maximum-subarray/submission-3.java
class Solution {
    public int maxSubArray(int[] nums) {
        int curr_sum=nums[0];
        int max_sum=nums[0];
        for(int i=1;i<nums.length;i++){
            curr_sum=Math.max(nums[i],nums[i]+curr_sum);
            max_sum=Math.max(max_sum,curr_sum);
        }
        return max_sum;
    }
}
